src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+}
+
+static int manager_new(Manager **ret) {
+        _cleanup_(manager_unrefp) Manager *m = NULL;
+        static const union sockaddr_union sa = {
+                .un.sun_family = AF_UNIX,
+                .un.sun_path = "/run/systemd/sysupdate/notify",
+        };
+        int r;
+
+        assert(ret);
+
+        m = new(Manager, 1);
+        if (!m)
+                return -ENOMEM;
+
+        m->targets = hashmap_new(&string_hash_ops);
+        if (!m->targets)
+                return -ENOMEM;
+
+        m->jobs = hashmap_new(&trivial_hash_ops);
+        if (!m->jobs)
+                return -ENOMEM;
+
+        r = sd_event_default(&m->event);
+        if (r < 0)
+                return r;
+
+        (void) sd_event_set_watchdog(m->event, true);
+
+        r = sd_event_add_signal(m->event, NULL, SIGINT, NULL, NULL);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_signal(m->event, NULL, SIGTERM, NULL, NULL);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_signal(m->event, NULL, SIGRTMIN+18, sigrtmin18_handler, NULL);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_memory_pressure(m->event, NULL, NULL, NULL);
+        if (r < 0)
+                log_debug_errno(r, "Failed allocate memory pressure event source, ignoring: %m");
+
+        r = sd_bus_default_system(&m->bus);
+        if (r < 0)
+                return r;
+
+        m->notify_fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (m->notify_fd < 0)
+                return -errno;
+
+        (void) mkdir_parents_label(sa.un.sun_path, 0755);
+        (void) sockaddr_un_unlink(&sa.un);
+
+        if (bind(m->notify_fd, &sa.sa, SOCKADDR_UN_LEN(sa.un)) < 0)
+                return -errno;
+
+        r = setsockopt_int(m->notify_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_io(m->event, &m->notify_event_source, m->notify_fd, EPOLLIN,
+                            manager_on_notify, m);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+static int manager_discover_images(Manager *m, TargetClass class) {
+        _cleanup_hashmap_free_ Hashmap *images = NULL;
+        Image *image;
+        int r;
+
+        images = hashmap_new(&image_hash_ops);
+        if (!images)
+                return log_oom();
+
+        r = image_discover((ImageClass) class, NULL, images);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enumerate images: %m");
+
+        HASHMAP_FOREACH(image, images) {
+                _cleanup_(target_unrefp) Target *t = NULL;
+
+                r = target_new(class, image->name, m, &t);
+                if (r < 0)
+                        return r;
+
+                t->path = strdup(image->path);
+                if (!t->path)
+                        return log_oom();
+
+                t->image_type = image->type;
```

Comment:
why not just keep a ref to the image object itself?

Reply:
I wanted to discourage using properties from the image, since some targets aren't backed by images (i.e. the host system and sysupdate components)

Reply:
dunno, internally sounds totally OK to simply pin the original image object if we have one, and set the pointer for it NULL otherwise. 

And then maybe add an accessor helper function for relevant fields that for each target that has an image object returns its internal field, otherwise the field on the target object. we do something like that at various other places actually.

but dunno, doesn't matter much

Reply:
Yeah I don't think it really matters either way, and doing the image+accessor increases the complexity a little bit. If you really want me to do it I will but for now I'll resolve this thread.